src/model.ts

```diff
@@ -1,4 +1,68 @@
 import WebSocket = require("ws");
+
+interface TimerBase {
+  isRunning: boolean;
+  timeLeft: number;
+}
+
+class Timer implements TimerBase {
+  isRunning: boolean;
+  #startTime: number;
+  timeLeft: number;
+  #intervalID;
+
+  constructor(allotedTime: number) {
+    this.isRunning = false;
+    this.#startTime = 0;
+    this.timeLeft = allotedTime;
+  }
+
+  #getTimeElapsedSinceLastStart() {
+    if (!this.#startTime) {
+      return 0;
+    }
+
+    return Date.now() - this.#startTime;
+  }
+  start() {
+    if (this.isRunning) {
+      throw "Timer is already running";
+    }
+    this.isRunning = true;
+    this.#startTime = Date.now();
+
+    this.#intervalID = setInterval(() => {
+      this.timeLeft = this.timeLeft - this.#getTimeElapsedSinceLastStart();
+      this.#startTime = Date.now();
+      console.log(this.timeLeft);
+      if (this.timeLeft <= 0) {
+        this.stop();
+        throw "Player Timed Out";
```

Comment:
We need to do a couple of things here:

1. In MessageTypes we have to add another message called `PLAYER_TIMEOUT`
2. We have to throw an object from here and not a string, the object should have `playerId`, `gameId`, `type: PLAYER_TIMEOUT`
3. In `server.js` file in the try catch handler, we have to check if the message type is `PLAYER_TIMEOUT`, if yes we have to call `engine.play(playerTimeoutMessage).then(log)` in the catch statement (for reference see how we handle player disconnect`
4. Handle the `PLAYER_TIMEOUT` message in `TiciTacaToeyGameEngine.ts` in all the method i.e. `validate`, `transition` and `notify`